nbt-utils/nbt_utils.cpp

```diff
@@ -270,6 +276,422 @@ wr_payload(LongArrayTag) {
+
+rd_snbt(DoubleTag) { stream >> value; return !!stream; }
+wr_snbt(DoubleTag) { stream.setf(std::ios::floatfield, std::ios::floatfield); stream.setf(std::ios::showpoint); stream << std::setprecision(std::numeric_limits<decltype(value)>::max_digits10) << value; }
+
+#undef rd_snbt_suffix
+
+#define rd_snbt_typed_array(elementType, typeMarker) \
+  rd_snbt(elementType##ArrayTag) { \
+    skip_whitespaces(stream); \
+    char c; \
+    stream.get(c); \
+    if (c != '[') \
+      return false; \
+    stream.get(c); \
+    if (c != typeMarker) \
+      return false; \
+    stream.get(c); \
+    if (c != ';') \
+      return false; \
+    const TagTypeMask acceptedMask = tagTypeToMask(TagType::elementType); \
+    elementType##Tag numTag; \
+    std::vector<decltype(numTag.value)> elements; \
+    while (stream) { \
+      skip_whitespaces(stream); \
+      if (!stream) \
+        return false; \
+      if (stream.peek() == ']') { \
+        stream.ignore(); \
+        break; \
+      } \
+      if (!(getNextSNBTTagTypes(stream) & acceptedMask)) \
+        return false; \
+      if(!numTag.readSNBT(stream)) \
+        return false; \
+      elements.push_back(numTag.value); \
+      skip_whitespaces(stream); \
+      stream.get(c); \
+      if (c == ']') \
+        break; \
+      if (c != ',') \
+        return false; \
+    } \
+    value.resize(elements.size()); \
+    for (int i = 0; i < elements.size(); ++i) \
+      value.setElement(i, elements[i]); \
+    return true; \
+  }
+
+rd_snbt_typed_array(Byte, 'B')
+wr_snbt(ByteArrayTag) {
+  stream << "[B;";
+  ByteTag numTag;
+  uint32_t count = htonl(value.count);
+  uint8_t* ptr = value.data.get();
+  for (uint32_t i = 0; i < count; ++i) {
+    numTag.value = ptr[i];
+    numTag.writeSNBT(stream);
+    if (i < count - 1)
+      stream << ',';
+  }
+  stream << ']';
+}
+
+rd_snbt(StringTag) {
+  skip_whitespaces(stream);
+  std::set<char> stringEnd;
+  switch (stream.peek()) {
+    case '"':
+    case '\'':
+      char c;
+      stream.get(c);
+      stringEnd = {c};
+      break;
+    default:
+      stringEnd = {' ', ',', ':', ']', '}'};
+      break;
+  }
+  std::vector<char> characters;
+  while (stream) {
+    auto c = stream.peek();
+    if (c == '\\') {
+      stream.ignore();
+      c = stream.peek();
+    } else {
+      if (stringEnd.count(c)) {
+        value = std::string(characters.begin(), characters.end());
+        if (c == '\'' || c == '"')
+          stream.ignore();
+        return true;
+      }
+    }
+    characters.push_back(c);
+    stream.ignore();
+  }
+  return false;
+}
+wr_snbt(StringTag) {
+  stream << '"';
+  uint16_t count = htons(value.length());
+  for (auto it = value.begin(); it != value.end(); ++it) {
+    char c = *it;
+    switch (c) {
+      case '"':
+      case '\'':
```

Comment:
Do we really need to escape single quotes given that your output is always encapsulated by double quotes?

Reply:
No, but I feel like it is less likely to cause problems if somebody would write a bad parser

Reply:
@Vftdan the problem is that Minecraft will emit the error `Invalid escape sequence \'` when you escape single quotes.